src/react/Reconciler.js

```diff
@@ -0,0 +1,293 @@
+      }
+      // we expect that if there's a context argument, it should always have Flow annotations
+      invariant(
+        componentType.$FormalParameters[1].typeAnnotation,
+        `__registerReactComponentRoot() failed due to root component missing Flow type annotations for the "context" argument`
+      );
+      contextTypes = flowAnnotationToObjectTypeTemplate(componentType.$FormalParameters[1].typeAnnotation);
+    }
+    return this.realm.wrapInGlobalEnv(() =>
+      this.realm.evaluateForEffects(() => {
+        let initialProps = createAbstractObject(this.realm, propsName, propTypes);
+        let initialContext = createAbstractObject(this.realm, contextName, contextTypes);
+        try {
+          let { result } = this._renderAsDeepAsPossible(
+            componentType,
+            initialProps,
+            initialContext,
+            BranchStatus.NO_BRANCH
+          );
+          this.statistics.optimizedTrees++;
+          return result;
+        } catch (e) {
+          // if there was a bail-out on the root component in this reconcilation process, then this
+          // should be an invariant as the user has explicitly asked for this component to get folded
+          let diagnostic = new CompilerDiagnostic(
+            "__registerReactComponentRoot() failed due to root component bailing out",
+            this.realm.currentLocation,
+            "PP0019",
+            "FatalError"
+          );
+          this.realm.handleError(diagnostic);
+          throw new FatalError();
+        }
+      })
+    );
+  }
+  _renderAsDeepAsPossible(
+    componentType: ECMAScriptSourceFunctionValue,
+    props: ObjectValue | AbstractValue,
+    context: ObjectValue | AbstractValue,
+    branchStatus: BranchStatusEnum
+  ) {
+    let { value, commitDidMountPhase, childContext } = this._renderOneLevel(
+      componentType,
+      props,
+      context,
+      branchStatus
+    );
+    let result = this._resolveDeeply(value, childContext, branchStatus);
+    return {
+      result,
+      childContext,
+      commitDidMountPhase,
+    };
+  }
+  _renderOneLevel(
+    componentType: ECMAScriptSourceFunctionValue,
+    props: ObjectValue | AbstractValue,
+    context: ObjectValue | AbstractValue,
+    branchStatus: BranchStatusEnum
+  ) {
+    if (valueIsClassComponent(this.realm, componentType)) {
+      // for now we don't support class components, so we throw a ExpectedBailOut
+      throw new ExpectedBailOut("class components not yet supported");
+    } else {
+      invariant(componentType.$Call, "Expected componentType to be a FunctionValue with $Call method");
+      let value = componentType.$Call(this.realm.intrinsics.undefined, [props, context]);
+      return { value, commitDidMountPhase: null, childContext: context };
+    }
+  }
+  _applyBranchedLogic(value: ObjectValue) {
+    // we need to apply a key when we're branched
+    let currentKeyValue = Get(this.realm, value, "key") || this.realm.intrinsics.null;
```

Comment:
Calling `Get` in this way can execute arbitrary code. Is that okay?

Reply:
Yep, this should be fine.

Reply:
This is fine, an empty string will be appended to the new string which is never empty.